b2luigi/batch/processes/gbasf2.py

```diff
@@ -0,0 +1,764 @@
+        n_jobs_by_status = Counter()
+        for _, job_info in job_status_dict.items():
+            n_jobs_by_status[job_info["Status"]] += 1
+
+        # print summary of jobs in project if setting is set and job status changed
+        if (get_setting("gbasf2_print_status_updates", default=True, task=self.task) and
+                n_jobs_by_status != self._n_jobs_by_status):
+            time_string = datetime.now().strftime("%Y-%m-%d %H:%M:%S")
+            job_status_string = str(dict(sorted(n_jobs_by_status.items()))).strip("{}")
+            print(f"Jobs in gbasf2 project \"{self.gbasf2_project_name}\" at {time_string}: {job_status_string}")
+        self._n_jobs_by_status = n_jobs_by_status
+
+        n_jobs = len(job_status_dict)
+        n_done = n_jobs_by_status["Done"]
+        # note: "Killed" jobs also get moved to "Failed" after a while
+        n_failed = n_jobs_by_status["Failed"]
+        n_in_final_state = n_done + n_failed
+
+        # The gbasf2 project is considered as failed if any of the jobs in it failed.
+        # However, we first try to reschedule thos jobs and only declare it as failed if the maximum number of retries
+        # for reschedulinhas been reached
+        if n_failed > 0:
+            self._on_failure_action()
+            if self.max_retries > 0 and self._reschedule_failed_jobs():
+                return JobStatus.running
+            return JobStatus.aborted
+
+        if n_in_final_state < n_jobs:
+            return JobStatus.running
+
+        # Require all jobs to be done for project success, any job failure results in a failed project
+        if n_done == n_jobs:
+            # download dataset only the first time that we return JobStatus.successful
+            if not self._project_had_been_successful:
+                self._on_first_success_action()
+                self._project_had_been_successful = True
+            return JobStatus.successful
+
+        raise RuntimeError("Could not determine JobStatus")
+
+    def _on_first_success_action(self):
+        """
+        Things to do after all jobs in the project had been successful, e.g. downloading the dataset and logs
+        """
+        self._download_logs()
+        if get_setting("gbasf2_download_dataset", default=True, task=self.task):
+            self._download_dataset()
+
+    def _on_failure_action(self):
+        """
+        Things to do after the project failed
+        """
+        job_status_dict = get_gbasf2_project_job_status_dict(self.gbasf2_project_name, self.gbasf2_env, dirac_user=self.dirac_user)
+        failed_job_dict = {job_id: job_info for job_id, job_info in job_status_dict.items()
+                           if job_info["Status"] == "Failed"}
+        n_failed = len(failed_job_dict)
+        print(f"{n_failed} failed jobs:\n{failed_job_dict}")
+        self._download_logs()
+
+    def _reschedule_failed_jobs(self):
```

Comment:
I do like your idea. However, luigi already comes with a `retry` configuration (https://luigi.readthedocs.io/en/stable/configuration.html#per-task-retry-policy). Would this also work? I know, you implement a lot more functionality (e.g. the storage to files, which would only be possible by using a central scheduler in the luigi-only case) - but of course less own code is always better.

Reply:
I retry single (sub-) jobs in a gbasf2 project and keep count of the number of retries for each individual jobs. The reason is that I regard a project a failed if a single job in it failed and then I want to only retry that job. If 100 jobs in a project fail, the global task/project retry counter would tell me that the task had 100 retries. Another possiblity would be to wait for all jobs to finish running, so they are either done or failed, and then retry all jobs in the project in one go (the `gb2_job_reschedule` command can send a retry command to all failed jobs in a project).

Actually, now that I think about it, having a cumulative per-job retry counter is not that bad, b/c if you have 100 failed jobs the user should look at his jobs, because then it's unlikely to be an issue with the grid site.

Reply:
I just recently found out that gbasf2 sometimes also automatically reschedules task. I didn't know about it until that happened to Max when testing. It seems to be documented in the Gbas2 FAQ on confluence (yeah...): Q: Are failed jobs automatically rescheduled?

> A: Jobs that failed for certain patterns are rescheduled automatically. Others are not.
> Q: How I can see if my jobs have been rescheduled?
> A: The "Job Monitor" on the DIRAC web portal does not show the counter by default, but one can choose to see it via "columns" option.

I don't know what the "certain patterns" are (I might ask today in b2questions). At least Max says that it does not reschedule all jobs that failed where it wasn't the jobs' failure, as some jobs only succeed after rescheduling manually. Gbasf2 has also a maximum number of retries/reschedules, I didn't find it anywhere documented but I grepped through the code and found a `num_reschedules=10`, so that might be it. You can see the current number of reschedules in the web gui as the last answer in the FAQ excerpt above states. I havent' found out how to find that out via the CLI.

But having such functionality in gbasf2 actually speaks against re-implementing it in the wrapper. E.g. my retry-counter doesn't count the automatic reschedules done by gbasf2. So that's a good reason to remove or at least simplify my retry code e.g. by building on the existing functionality.

I just have to think how to achieve that when the task is retried via the luigi retry policy, how to run the rescheduling command instead of `start_job`, the rest is easy.